src/classes/MinisignKeys.php

```diff
@@ -0,0 +1,192 @@
+use SensitiveParameter;
+
+use function hash_equals;
+use function preg_match;
+use function sodium_crypto_generichash;
+use function sodium_memzero;
+use function unpack;
+
+use const SODIUM_CRYPTO_SIGN_PUBLICKEYBYTES;
+use const SODIUM_CRYPTO_SIGN_SECRETKEYBYTES;
+
+/**
+ * This class is a representation of a minisign key
+ * minisign by Frank Denis: https://jedisct1.github.io/minisign/#secret-key-format
+ * Inspired by the code from https://github.com/soatok/minisign-php by Soatok Dreamseeker
+ */
+readonly class MinisignKeys
+{
+    // ed25519
+    private const string SIGNATURE_ALGO = 'Ed';
+
+    private const int KEYID_BYTES = 8;
+
+    private const int KDF_LENGTH = 104;
+
+    // our key derivation algo: scrypt
+    private const string KDF_ALGO = 'Sc';
+
+    // our checksum algo: blake2
+    private const string CKSUM_ALGO = 'B2';
+
+    public function __construct(
+        public string $signatureAlgo,
+        public string $id,
+        /** @var non-empty-string */
+        public string $priv,
+        public string $pub,
+        public string $salt,
+        public string $derivedKey,
+    ) {}
+
+    public function getIdHex(): string
+    {
+        return Hex::encode($this->id);
+    }
+
+    public static function generate(#[SensitiveParameter] string $passphrase): self
+    {
+        // Generate a salt for key derivation
+        // SCRYPT_SALSA208SHA256 salt should be crypto_pwhash_scryptsalsa208sha256_SALTBYTES bytes
+        $salt = random_bytes(SODIUM_CRYPTO_PWHASH_SCRYPTSALSA208SHA256_SALTBYTES);
+
+        // derive a key from the passphrase
+        $derivedKey = self::kdf(
+            $passphrase,
+            $salt,
+            SODIUM_CRYPTO_PWHASH_SCRYPTSALSA208SHA256_OPSLIMIT_INTERACTIVE,
+            SODIUM_CRYPTO_PWHASH_SCRYPTSALSA208SHA256_MEMLIMIT_INTERACTIVE,
+        );
+
+        // the key ID is 8 random bytes to give a hint about which secret key was used to sign a message
+        $id = random_bytes(self::KEYID_BYTES);
+
+        // generate a random Ed25519 keypair as one string
+        $keypair = sodium_crypto_sign_keypair();
+        $pub = sodium_crypto_sign_publickey($keypair);
+        /** @var non-empty-string */
+        $priv = sodium_crypto_sign_secretkey($keypair);
+
+        return new self(self::SIGNATURE_ALGO, $id, $priv, $pub, $salt, $derivedKey);
+    }
+
+    /**
+     * Secret key format from: https://jedisct1.github.io/minisign/#secret-key-format
+     * untrusted comment: <arbitrary text>
+     * base64(<signature_algorithm> || <kdf_algorithm> || <cksum_algorithm> || <kdf_salt> || <kdf_opslimit> || <kdf_memlimit> || <keynum_sk>)
+     */
+    public static function deserialize(string $secretKey, #[SensitiveParameter] string $passphrase): self
+    {
+        $sk = array();
+        if (!preg_match(SignatureHelper::REGEX, $secretKey, $sk)) {
+            throw new ImproperActionException(_('Invalid secret key format!'));
+        }
+        $decoded = Base64::decode($sk[2]);
```

Comment:
if `SignatureHelper::REGEX` is changed (see other comment) adjust array index to 1